File: game.of.life.v3/game.of.life.v3/Cycle.cs
namespace game.of.life.v3
{
    using System.Linq;

    public class Cycle
    {

        public RectangularInfinite2DGrid Run(RectangularInfinite2DGrid grid)
        {
            var newGrid = new RectangularInfinite2DGrid();
            newGrid.AddCells(grid.Cells.Select(c => new Cell(c.X, c.Y, c.State)).ToArray());
            newGrid.Discover();

            newGrid.Cells.ToList().ForEach(cell => cell.ComputeMutation(newGrid.GetNeighbours(cell).Count(c => c.IsAlive)));
            newGrid.Cells.ToList().ForEach(cell => cell.CompleteMutation());

            newGrid.Discover();
            newGrid.Clean();

            return newGrid;
        }
    }
}

File: game.of.life/game.of.life.ts.web/src/scripts/GoL/Logic/Cycle.ts
/// <reference path="../../../../typings/linq.d.ts"/>

'use strict';

namespace GoL.Logic {
	import Cell = Gol.Logic.Cell;

	export class Cycle {
		public Run(currentGrid: RectangularInfinite2DGrid): RectangularInfinite2DGrid {
			const newGrid: RectangularInfinite2DGrid = this.InitializeNewGrid(currentGrid);

			this.ComputeAndCompleteMutation(newGrid);

			this.PrepareForNextCycle(newGrid);

			return newGrid;
		}

		private InitializeNewGrid(currentGrid: RectangularInfinite2DGrid): RectangularInfinite2DGrid {
			const cells: Cell[] = Enumerable.from(currentGrid.cells).select((c: Cell) => new Cell(c.x, c.y, c.state)).toArray();
			const newGrid = new RectangularInfinite2DGrid(cells);
			newGrid.discover();
			return newGrid;
		}

		private ComputeAndCompleteMutation(newGrid: RectangularInfinite2DGrid): void {
			Enumerable.from(newGrid.cells).forEach((cell: Cell) => cell.computeNextMutation(Enumerable.from(newGrid.getNeighbours(cell)).count((c: Cell) => c.isAlive())));
			Enumerable.from(newGrid.cells).forEach((cell: Cell) => cell.completeMutation());
		}

		private PrepareForNextCycle(newGrid: RectangularInfinite2DGrid): void {
			newGrid.discover();
			newGrid.clean();
		}
	}
}
